kafka-ui-api/src/main/java/com/provectus/kafka/ui/emitter/BackwardRecordEmitter.java

```diff
@@ -17,36 +17,47 @@
 import org.apache.kafka.common.utils.Bytes;
 import reactor.core.publisher.FluxSink;
 
-@RequiredArgsConstructor
 @Log4j2
 public class BackwardRecordEmitter
-    implements java.util.function.Consumer<FluxSink<ConsumerRecord<Bytes, Bytes>>> {
-
-  private static final Duration POLL_TIMEOUT_MS = Duration.ofMillis(1000L);
+    extends AbstractEmitter
+    implements java.util.function.Consumer<FluxSink<TopicMessageEvent>> {
 
   private final Function<Map<String, Object>, KafkaConsumer<Bytes, Bytes>> consumerSupplier;
   private final OffsetsSeekBackward offsetsSeek;
 
+  public BackwardRecordEmitter(
+      Function<Map<String, Object>, KafkaConsumer<Bytes, Bytes>> consumerSupplier,
+      OffsetsSeekBackward offsetsSeek,
+      RecordSerDe recordDeserializer) {
+    super(recordDeserializer);
+    this.offsetsSeek = offsetsSeek;
+    this.consumerSupplier = consumerSupplier;
+  }
+
   @Override
-  public void accept(FluxSink<ConsumerRecord<Bytes, Bytes>> sink) {
+  public void accept(FluxSink<TopicMessageEvent> sink) {
     try (KafkaConsumer<Bytes, Bytes> configConsumer = consumerSupplier.apply(Map.of())) {
       final List<TopicPartition> requestedPartitions =
           offsetsSeek.getRequestedPartitions(configConsumer);
+      sendPhase(sink, "Request partitions");
       final int msgsPerPartition = offsetsSeek.msgsPerPartition(requestedPartitions.size());
       try (KafkaConsumer<Bytes, Bytes> consumer =
                consumerSupplier.apply(
                    Map.of(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, msgsPerPartition)
                )
       ) {
+        sendPhase(sink, "Created consumer");
         final Map<TopicPartition, Long> partitionsOffsets =
             offsetsSeek.getPartitionsOffsets(consumer);
+        sendPhase(sink, "Requested partitions offsets");
```

Comment:
on line 50 its better to use sorted map, since later we iterate over partitions and it is expected to have them in order, imho